File: node/src/CompressionConfiguration.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 */


/**
 * Enum representing the compression backend to use for automatic compression.
 */
export enum CompressionBackend {
    /**
     * Use zstd compression backend.
     */
    ZSTD = "ZSTD",
    /**
     * Use lz4 compression backend.
     */
    LZ4 = "LZ4",
}

/**
 * Configuration for automatic compression of values.
 * 
 * @remarks
 * This configuration enables transparent compression and decompression of values
 * for set-type and get-type commands respectively. When enabled, values will be
 * automatically compressed before being sent to the server and decompressed when
 * received from the server.
 * 
 * The compression is completely transparent to the application layer and maintains
 * full backward compatibility with existing data and non-compression clients.
 * 
 * @example
 * ```typescript
 * const compressionConfig: CompressionConfiguration = {
 *   enabled: true,
 *   backend: CompressionBackend.ZSTD,
 *   compressionLevel: 3,
 *   minCompressionSize: 64,
 *   maxCompressionSize: 1024 * 1024, // 1MB
 * };
 * ```
 */
export interface CompressionConfiguration {
    /**
     * Whether compression is enabled.
     * @default false
     */
    enabled: boolean;

    /**
     * The compression backend to use.
     * @default CompressionBackend.ZSTD
     */
    backend: CompressionBackend;

    /**
     * The compression level to use. If not set, the backend's default level will be used.
     * 
     * Valid ranges:
     * - ZSTD: 1-22 (higher values provide better compression but slower speed)
     * - LZ4: 1-12 (higher values provide better compression but slower speed)
     * 
     * @default undefined (uses backend default)
     */
    compressionLevel?: number;

    /**
     * The minimum size in bytes for values to be compressed.
     * Values smaller than this will not be compressed to avoid overhead.
     * @default 64
     */
    minCompressionSize: number;

    /**
     * The maximum size in bytes for values to be compressed.
     * Values larger than this will not be compressed.
     * If not set, no maximum limit is applied.
     * @default undefined (no limit)
     */
    maxCompressionSize?: number;
}

/**
 * Validates compression configuration parameters.
 * 
 * @param config - The compression configuration to validate
 * @throws {ConfigurationError} If any configuration parameter is invalid
 * 
 * @example
 * ```typescript
 * const config: CompressionConfiguration = {
 *   enabled: true,
 *   backend: CompressionBackend.ZSTD,
 *   compressionLevel: 3,
 *   minCompressionSize: 64,
 *   maxCompressionSize: 1024,
 * };
 * 
 * validateCompressionConfiguration(config); // Throws if invalid
 * ```
 */
export function validateCompressionConfiguration(
    config: CompressionConfiguration,
): void {
    if (config.minCompressionSize < 0) {
        throw new ConfigurationError(
            "minCompressionSize must be non-negative",
        );
    }

    if (config.maxCompressionSize !== undefined) {
        if (config.maxCompressionSize < 0) {
            throw new ConfigurationError(
                "maxCompressionSize must be non-negative",
            );
        }
        if (config.maxCompressionSize < config.minCompressionSize) {
            throw new ConfigurationError(
                "maxCompressionSize must be greater than or equal to minCompressionSize",
            );
        }
    }

    if (config.compressionLevel !== undefined) {
        // Validate compression level based on backend
        if (config.backend === CompressionBackend.ZSTD) {
            if (config.compressionLevel < 1 || config.compressionLevel > 22) {
                throw new ConfigurationError(
                    "compressionLevel for ZSTD backend must be between 1 and 22",
                );
            }
        } else if (config.backend === CompressionBackend.LZ4) {
            if (config.compressionLevel < 1 || config.compressionLevel > 12) {
                throw new ConfigurationError(
                    "compressionLevel for LZ4 backend must be between 1 and 12",
                );
            }
        }
    }
}

/**
 * Creates a compression configuration with default values.
 * 
 * @param overrides - Optional overrides for default values
 * @returns A validated compression configuration
 * 
 * @example
 * ```typescript
 * // Create with defaults
 * const defaultConfig = createCompressionConfiguration();
 * 
 * // Create with overrides
 * const customConfig = createCompressionConfiguration({
 *   enabled: true,
 *   backend: CompressionBackend.LZ4,
 *   compressionLevel: 6,
 * });
 * ```
 */
export function createCompressionConfiguration(
    overrides: Partial<CompressionConfiguration> = {},
): CompressionConfiguration {
    const config: CompressionConfiguration = {
        enabled: false,
        backend: CompressionBackend.ZSTD,
        minCompressionSize: 64,
        ...overrides,
    };

    validateCompressionConfiguration(config);
    return config;
}

File: csharp/lib/CompressionConfiguration.cs
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

namespace Glide;

/// <summary>
/// Represents the compression configuration for automatic compression of values.
/// </summary>
/// <remarks>
/// When compression is enabled, values that meet the minimum size threshold will be
/// automatically compressed before being sent to the server and decompressed when retrieved.
/// Compression is transparent to the user — compressed and uncompressed clients can
/// interoperate seamlessly.
///
/// This is a record type providing structural equality, making it easy to compare
/// configurations in tests and application code.
/// </remarks>
/// <param name="Enabled">Whether compression is enabled.</param>
/// <param name="Backend">The compression backend to use. Defaults to <see cref="CompressionBackend.Zstd"/>.</param>
/// <param name="CompressionLevel">The compression level. If null, the backend default is used.
/// Valid ranges are backend-specific and validated by the Rust core.
/// ZSTD default is 3, LZ4 default is 0.</param>
/// <param name="MinCompressionSize">Minimum value size in bytes for compression. Defaults to 64.
/// Must be at least <see cref="MinCompressedSize"/> (6 bytes).</param>
/// <exception cref="ConfigurationError">Thrown when configuration parameters are invalid.</exception>
public record CompressionConfiguration(
    bool Enabled = false,
    CompressionBackend Backend = CompressionBackend.Zstd,
    int? CompressionLevel = null,
    uint MinCompressionSize = 64)
{
    /// <summary>
    /// Minimum allowed value for <see cref="MinCompressionSize"/>.
    /// This is the header size (5 bytes) + 1 byte of data = 6 bytes.
    /// </summary>
    /// <remarks>
    /// IMPORTANT: This must stay in sync with the Rust core constant <c>MIN_COMPRESSED_SIZE</c>
    /// defined in glide-core/src/compression.rs. If the header format changes in the Rust core,
    /// this value must be updated to match. Once FFI is fully wired, consider reading this
    /// value from the native library instead of hardcoding it.
    /// </remarks>
    public const uint MinCompressedSize = 6;

    // Validate on construction. Record primary constructors don't support throwing,
    // so we validate in the init block via a property trick isn't available — instead
    // we validate here and the compiler calls this after setting all properties.
    private bool _validated = Validate(Enabled, Backend, CompressionLevel, MinCompressionSize);

    private static bool Validate(bool enabled, CompressionBackend backend, int? compressionLevel, uint minCompressionSize)
    {
        if (minCompressionSize < MinCompressedSize)
        {
            // Note: error message uses snake_case intentionally for cross-language consistency
            // with Python and Java bindings (see python/glide-shared/glide_shared/config.py).
            throw new ConfigurationError(
                $"min_compression_size should be at least {MinCompressedSize} bytes");
        }

        if (!Enum.IsDefined(typeof(CompressionBackend), backend))
        {
            throw new ConfigurationError(
                $"Invalid compression Backend: {backend}");
        }

        // Note: compression_level validation is performed by the Rust core,
        // which uses the actual compression library's valid ranges.
        // This ensures the validation stays in sync with library updates.

        return true;
    }
}
